File: backend/app/routers/scenarios.py
from typing import Annotated, List

from fastapi import APIRouter, Depends

from app.core.common import parse_uuid
from app.models.scenario import (
    Scenario,
    ScenarioAddCustomer,
    ScenarioCreate,
    ScenarioRemoveCustomer,
    ScenarioStart,
    ScenarioUpdate,
    ScenarioUpdateHistory,
)
from app.models.scenario_customer import ScenarioCustomer, ScenarioCustomerUpdate
from app.models.scenario_session import ScenarioSession
from app.services.scenario_service import ScenarioService
from app.services.trainee_service import TraineeService

router = APIRouter(prefix="/scenarios", tags=["Scenarios"])


@router.get("/")
async def get_scenarios(
    scenario_service: Annotated[ScenarioService, Depends()],
) -> list[Scenario]:
    """Get all scenarios."""
    return await scenario_service.get_scenarios()


@router.post("/")
async def create_scenario(
    scenario_data: ScenarioCreate,
    scenario_service: Annotated[ScenarioService, Depends()],
) -> Scenario:
    """Create a new scenario."""
    return await scenario_service.create_scenario(scenario_data)


@router.get("/{scenario_id}")
async def get_scenario(
    scenario_id: str,
    scenario_service: Annotated[ScenarioService, Depends()],
) -> Scenario:
    """Get a scenario by ID."""
    return await scenario_service.get_scenario(scenario_id)


@router.get("/{scenario_id}/customers")
async def get_scenario_customers(
    scenario_id: str,
    scenario_service: Annotated[ScenarioService, Depends()],
) -> List[ScenarioCustomer]:
    """Get all customers for a scenario."""
    return await scenario_service.get_scenario_customers(scenario_id)


@router.put("/{scenario_id}")
async def update_scenario(
    scenario_id: str,
    scenario_data: ScenarioUpdate,
    scenario_service: Annotated[ScenarioService, Depends()],
) -> Scenario:
    """Update a scenario."""
    return await scenario_service.update_scenario(scenario_id, scenario_data)


@router.post("/{scenario_id}/customers")
async def add_customer_to_scenario(
    scenario_id: str,
    customer_data: ScenarioAddCustomer,
    scenario_service: Annotated[ScenarioService, Depends()],
) -> Scenario:
    """Add a customer to a scenario."""
    return await scenario_service.add_customer_to_scenario(scenario_id, customer_data)


@router.delete("/{scenario_id}/customers")
async def remove_customer_from_scenario(
    scenario_id: str,
    customer_data: ScenarioRemoveCustomer,
    scenario_service: Annotated[ScenarioService, Depends()],
) -> Scenario:
    """Remove a customer from a scenario."""
    return await scenario_service.remove_customer_from_scenario(
        scenario_id, customer_data
    )


@router.put("/{scenario_id}/history")
async def update_scenario_history(
    scenario_id: str,
    history_data: ScenarioUpdateHistory,
    scenario_service: Annotated[ScenarioService, Depends()],
) -> Scenario:
    """Update the history of a scenario."""
    return await scenario_service.update_scenario_history(scenario_id, history_data)


@router.post("/{scenario_id}/start")
async def start_scenario(
    scenario_id: str,
    scenario_data: ScenarioStart,
    trainee_service: Annotated[TraineeService, Depends()],
) -> ScenarioSession:
    """Start a scenario."""
    return await trainee_service.start_scenario(
        trainee_id=parse_uuid(scenario_data.trainee_id),
        scenario_id=parse_uuid(scenario_id),
    )


@router.put("/{scenario_id}/customers/{customer_id}")
async def update_scenario_customer(
    scenario_id: str,
    customer_id: str,
    customer_data: ScenarioCustomerUpdate,
    scenario_service: Annotated[ScenarioService, Depends()],
) -> ScenarioCustomer:
    """Update a customer in a scenario."""
    return await scenario_service.update_scenario_customer(
        scenario_id, customer_id, customer_data
    )


@router.delete("/{scenario_id}")
async def delete_scenario(
    scenario_id: str,
    scenario_service: Annotated[ScenarioService, Depends()],
) -> None:
    """Delete a scenario."""
    return await scenario_service.delete_scenario(scenario_id)

File: frontend/lib/api/scenarios.ts
import { fetchApi } from '@/lib/api/common';
import { Scenario, ScenarioCreate, ScenarioUpdate } from '@/types/scenario';
import { ScenarioCustomer, ScenarioCustomerUpdate } from '@/types/scenario-customer';
import { UserScenarioSession } from '@/types/user-scenario-session';

/**
 * Fetch all scenarios, optionally filtered by scheme ID or slug
 * @param schemeIdOrSlug - The scheme's ID or slug to filter by
 */
export async function getScenarios(schemeIdOrSlug?: string): Promise<Scenario[]> {
  if (schemeIdOrSlug) {
    return fetchApi<Scenario[]>(`/schemes/${schemeIdOrSlug}/scenarios`);
  }
  return fetchApi<Scenario[]>('/scenarios');
}

/**
 * Fetch a single scenario by ID
 * @param scenarioId - The scenario's ID
 */
export async function getScenario(scenarioId: string): Promise<Scenario> {
  return fetchApi<Scenario>(`/scenarios/${scenarioId}`);
}

/**
 * Create a new scenario
 * @param scenarioData - The scenario data to create
 */
export async function createScenario(
  scenarioData: ScenarioCreate,
): Promise<Scenario> {
  return fetchApi<Scenario>('/scenarios', {
    method: 'POST',
    body: JSON.stringify(scenarioData),
  });
}

/**
 * Update a scenario
 * @param scenarioId - The scenario's ID
 * @param scenarioData - The scenario data to update
 */
export async function updateScenario(
  scenarioId: string,
  scenarioData: ScenarioUpdate,
): Promise<Scenario> {
  return fetchApi<Scenario>(`/scenarios/${scenarioId}`, {
    method: 'PUT',
    body: JSON.stringify(scenarioData),
  });
}

/**
 * Delete a scenario
 * @param scenarioId - The scenario's ID to delete
 */
export async function deleteScenario(scenarioId: string): Promise<void> {
  return fetchApi<void>(`/scenarios/${scenarioId}`, {
    method: 'DELETE',
  });
}

/**
 * Start a scenario
 * @param scenarioId - The scenario's ID
 * @param traineeId - The trainee's ID
 */
export async function startScenario(
  scenarioId: string,
  traineeId: string,
): Promise<UserScenarioSession> {
  return fetchApi<UserScenarioSession>(`/scenarios/${scenarioId}/start`, {
    method: 'POST',
    body: JSON.stringify({ trainee_id: traineeId }),
  });
}

/**
 * Get the active scenario session for a trainee
 * @param traineeId - The trainee's ID
 */
export async function getActiveScenarioSession(
  traineeId: string,
): Promise<UserScenarioSession> {
  return fetchApi<UserScenarioSession>(`/scenarios/active`, {
    method: 'GET',
    body: JSON.stringify({ trainee_id: traineeId }),
  });
}

/**
 * Get all customers for a scenario
 * @param scenarioId - The scenario's ID
 */
export async function getScenarioCustomers(
  scenarioId: string,
): Promise<ScenarioCustomer[]> {
  return fetchApi<ScenarioCustomer[]>(`/scenarios/${scenarioId}/customers`);
}

/**
 * Add a customer to a scenario
 * @param scenarioId - The scenario's ID
 * @param customerId - The customer's ID to add
 * @param name - Optional custom name for the customer
 */
export async function addCustomerToScenario(
  scenarioId: string,
  customerId: string,
  name?: string,
): Promise<Scenario> {
  return fetchApi<Scenario>(`/scenarios/${scenarioId}/customers`, {
    method: 'POST',
    body: JSON.stringify({
      customer_id: customerId,
      name: name
    }),
  });
}

/**
 * Remove a customer from a scenario
 * @param scenarioId - The scenario's ID
 * @param customerId - The customer's ID to remove
 * @param scenarioCustomerId - The specific scenario customer ID to remove
 */
export async function removeCustomerFromScenario(
  scenarioId: string,
  customerId: string,
  scenarioCustomerId?: string,
): Promise<Scenario> {
  return fetchApi<Scenario>(`/scenarios/${scenarioId}/customers`, {
    method: 'DELETE',
    body: JSON.stringify({
      customer_id: customerId,
      id: scenarioCustomerId
    }),
  });
}

/**
 * Update a customer in a scenario
 * @param scenarioId - The scenario's ID
 * @param customerId - The customer's ID
 * @param customerData - The customer data to update
 */
export async function updateScenarioCustomer(
  scenarioId: string,
  customerId: string,
  customerData: ScenarioCustomerUpdate,
): Promise<ScenarioCustomer> {
  return fetchApi<ScenarioCustomer>(`/scenarios/${scenarioId}/customers/${customerId}`, {
    method: 'PUT',
    body: JSON.stringify(customerData),
  });
}
